### ISSUE TITLE
Quick Question about BigDecimal and If function

### ISSUE THREAD

**Issue by fobbyal** [uklimaschewski/EvalEx#62]
The project I am using EvalEx on is money based therefore all calculation falls in the limits of double. BigDecimal arithmetic function is known to be much more slower than primitive like double. I am exploring the possibility of wrapping the Values EvalEX in a pseudo class that supports BigDecimal as well as primitives like double or even floats in extreme case. For use cases that are can tolerate precision losses based on performance gain.

This is going to dramatically change the code so i am asking to see if you will be willing to accept such a pull request.

If i go ahead with this modification. I would like to improve the if function as well. Consider the following
`if(a =0, 0 , 12/a)`
The above does not work well because 12/a is evaluated regardless of the outcome of `a=0`. This was a problem i ran into. I solved it by returning 0 for any divisible by 0 in my specific context. This is not ideal. I would like to lazily evaluate the parameters of the if based on the outcome of the boolean tester.

**Comment by uklimaschewski** [comment]
If you are working with money, `BigDecimal` should be the way to go. Unless you really need the performance of `double`. With `double`, you have to take care for proper rounding in each part of the expression, else you will end up with missing cents everywhere.
That was one of the original reasons, why I created this expression parser. All other parsers around were using `double` and were therefore unusable in a financial environment, where I needed it.

Have a look at this:
````java
    System.out.println( "362.2 - 362.6 = " + ( 362.2 - 362.6 ) );
````
it will print
````
    362.2 - 362.6 = -0.4000000000000341
````
Which is horrific, if you use this as part of a sum in a financial environment without rounding even the substraction.
You will probably not notice any problem, until the sum of the resulting doubles suddenly end up being a cent to high or low.

If you need the performance, `double` is much faster, but it will only be noticeable when calculating millions of expressions.

have a look at [this blog article](http://java-performance.info/bigdecimal-vs-double-in-financial-calculations/), which sums up the pros and cons of `BigDecimal` vs. `double` in monetary calculations.

Somebody did such a modification to EvalEx a while back, but it ended up in a more complex library than I wanted to maintain. I would like to keep this small and compact, within only one class.

The problem with the `IF()` function is a constant source of misunderstanding. Let me see how you plan to 'fix' this issue, I would be happy to include any good solution.

**Comment by fobbyal** [comment]
Here is my plan for the LazyIf..
I plan to change the `public BigDecimal eval()` so that it uses a `Stack<LazyNumber>` instad of `Stack<BigDecimal>` for processing the RPN. See changes below (_Sorry for doing it this way. I am prototyping and my IDE keeps fighting the project format. It will not look nice if i pushed the changes. I just want to run the idea by you so we are the same page. If you are okay with it then I'll spend the time to add tests and format the code based on the project's preferences._`). Let me know your thoughts.

* New Interface
```
interface LazyNumber {
    BigDecimal eval();
}
```

* New method in Function class All lazy featured functions are suppose to override this method instead of `eval(List<BigDecimal>)` This is the done this way to prevent breaking changes.

```
		public LazyNumber lazyEval(List<LazyNumber> lazyParams) {
			final List<BigDecimal> bigDecimalParams = new ArrayList<BigDecimal>();
			for (LazyNumber num : lazyParams) {
				bigDecimalParams.add(num.eval());
			}
			return new LazyNumber() {
				public BigDecimal eval() {
					return Function.this.eval(bigDecimalParams);
				}
			};
		}
```



* Variables
```
				LazyNumber number = new LazyNumber() {
					public BigDecimal eval() {
						return variables.get(token).round(mc);
					}
				};
				stack.push(number);
```
* Operators
```
				final LazyNumber v1 = stack.pop();
				final LazyNumber v2 = stack.pop();
				LazyNumber number = new LazyNumber() {
					public BigDecimal eval() {
						return operators.get(token).eval(v2.eval(), v1.eval());
					}
				};
				stack.push(number);
```
* Constants
```
				stack.push(new LazyNumber() {
					public BigDecimal eval() {
						return new BigDecimal(token, mc);
					}
				});
```
* Functions
```
				Function f = functions.get(token.toUpperCase(Locale.ROOT));
				ArrayList<LazyNumber> p = new ArrayList<LazyNumber>(
						!f.numParamsVaries() ? f.getNumParams() : 0);
				// pop parameters off the stack until we hit the start of
				// this function's parameter list
				while (!stack.isEmpty() && stack.peek() != PARAMS_START) {
					p.add(0, stack.pop());
				}
				if (stack.peek() == PARAMS_START) {
					stack.pop();
				}
				if (!f.numParamsVaries() && p.size() != f.getNumParams()) {
					throw new ExpressionException("Function " + token + " expected " + f.getNumParams() + " parameters, got " + p.size());
				}
				stack.push(f.lazyEval(p));
```

**Comment by fobbyal** [comment]
#63

**Comment by uklimaschewski** [comment]
Thank you and sorry for the delay in answering!
Looking at the code and your (modified) example:

    IF(a=0, a+1, 12/a)

From what I see, your code changes alone do not solve the problem of `12/a` and `a+1` being evaluated prior to returning the `IF` result and therefore still causing a division by zero.
I have quiickly tested it, but maybe I'm missing an important part?

**Comment by fobbyal** [comment]
This was just a explanation not the fully implemented code. The fully implemented code would have to change the if function to a lazy function. Please take a look at my pull request #63 it has one test of `if(a=0,0,12/a)` that is working.
Assuming that a is 0
`if(a=0,0,12/a)` will evaluate 0
`if(a=0,a+1,12/a)` will evaluate 1 

sorry for the confusion

**Comment by fobbyal** [comment]
bump

**Comment by uklimaschewski** [comment]
Thank you, I have merged your pull request #63.
I was thinking of an easier way to do this, but didn't come to a better solution.
Another solution might have been the possibility of using Strings instead of BigDecimal as function parameters and then passing the three expression strings of the `IF` function without evaluating them.
But this would have led to other problems.
So I'll give this a try.

**Comment by fobbyal** [comment]
thx.. great...
